server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java

```diff
@@ -0,0 +1,523 @@
+            builderReference.set(MetaData.builder(metaData));
+        });
+
+        final MetaData.Builder builder = builderReference.get();
+        assert builder != null : "no global metadata found";
+
+        logger.trace("got global metadata, now reading index metadata");
+
+        final Set<String> indexUUIDsForAssertions = new HashSet<>();
+        consumeFromType(searcher, INDEX_TYPE_NAME, bytes ->
+        {
+            final IndexMetaData indexMetaData = IndexMetaData.fromXContent(XContentFactory.xContent(XContentType.SMILE)
+                .createParser(namedXContentRegistry, LoggingDeprecationHandler.INSTANCE, bytes));
+            logger.trace("found index metadata for {}", indexMetaData.getIndex());
+            //noinspection AssertWithSideEffects
+            assert indexUUIDsForAssertions.add(indexMetaData.getIndexUUID());
+            builder.put(indexMetaData, false);
+        });
+
+        final Map<String, String> userData = reader.getIndexCommit().getUserData();
+        logger.trace("loaded metadata [{}] from [{}]", userData, reader.directory());
+        assert userData.size() == 4 : userData;
+        assert userData.get(CURRENT_TERM_KEY) != null;
+        assert userData.get(LAST_ACCEPTED_VERSION_KEY) != null;
+        assert userData.get(NODE_ID_KEY) != null;
+        assert userData.get(NODE_VERSION_KEY) != null;
+        return new OnDiskState(userData.get(NODE_ID_KEY), Long.parseLong(userData.get(CURRENT_TERM_KEY)),
+            Long.parseLong(userData.get(LAST_ACCEPTED_VERSION_KEY)), builder.build());
+    }
+
+    private static void consumeFromType(IndexSearcher indexSearcher, String type,
+                                        CheckedConsumer<byte[], IOException> docValuesConsumer)
+        throws IOException {
+
+        final Query query = new TermQuery(new Term(TYPE_FIELD_NAME, type));
+        final Weight weight = indexSearcher.createWeight(query, ScoreMode.COMPLETE_NO_SCORES, 0.0f);
+        logger.trace("running query [{}]", query);
+
+        for (LeafReaderContext leafReaderContext : indexSearcher.getIndexReader().leaves()) {
+            logger.trace("new leafReaderContext: {}", leafReaderContext);
+            final Scorer scorer = weight.scorer(leafReaderContext);
+            final Bits liveDocs = leafReaderContext.reader().getLiveDocs();
+            final IntPredicate isLiveDoc = liveDocs == null ? i -> true : liveDocs::get;
+            if (scorer != null) {
+                final DocIdSetIterator docIdSetIterator = scorer.iterator();
+                while (docIdSetIterator.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+                    if (isLiveDoc.test(docIdSetIterator.docID())) {
+                        logger.trace("processing doc {}", docIdSetIterator.docID());
+                        docValuesConsumer.accept(
+                            indexSearcher.getIndexReader().document(docIdSetIterator.docID()).getBinaryValue(DATA_FIELD_NAME).bytes);
```

Comment:
This will be wrong if you have multiple segments as docID is a segment doc ID here and you are using it against a top-level reader.

```suggestion
                            leafReaderContext.reader().document(docIdSetIterator.docID()).getBinaryValue(DATA_FIELD_NAME).bytes);
```

Reply:
Thanks, well spotted, addressed in 1027c8c56b91fcb30275f73172d66ab7cb83eed1 along with a test.